core/src/main/java/org/geysermc/geyser/item/GeyserCustomItemManager.java

```diff
@@ -0,0 +1,181 @@
+import org.jetbrains.annotations.NotNull;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class GeyserCustomItemManager extends CustomItemManager {
+    public static final String CUSTOM_PREFIX = "geyser_custom:";
+
+    private Map<String, List<CustomItemData>> customMappings = new HashMap<>();
+
+    private List<GeyserCustomMappingData> registeredCustomItems = new ArrayList<>();
+    private Int2ObjectMap<List<ComponentItemData>> registeredComponentItems = new Int2ObjectOpenHashMap<>();
+
+    public GeyserCustomItemManager() {
+        MappingsConfigReader.init(this);
+    }
+
+    public void loadMappingsFromJson() {
+        Path customMappingsDirectory = MappingsConfigReader.getCustomMappingsDirectory();
+        if (!Files.exists(customMappingsDirectory)) {
+            try {
+                Files.createDirectories(customMappingsDirectory);
+            } catch (IOException e) {
+                GeyserImpl.getInstance().getLogger().error("Failed to create custom mappings directory", e);
+                return;
+            }
+        }
+
+        Path[] mappingsFiles = MappingsConfigReader.getCustomMappingsFiles();
+        for (Path mappingsFile : mappingsFiles) {
+            this.loadMappingsFromJson(mappingsFile);
+        }
+
+        if (this.registeredItemCount() != 0) {
+            GeyserImpl.getInstance().getLogger().info("Registered " + this.registeredItemCount() + " custom items from mappings");
+        }
+    }
+
+    public void loadMappingsFromJson(@NotNull Path file) {
+        MappingsConfigReader.readMappingsFromJson(file);
+    }
+
+    private boolean cantRegisterNewItem() {
+        return GeyserImpl.getInstance().isInitialized() || !GeyserImpl.getInstance().getConfig().isAddNonBedrockItems();
+    }
+
+    @Override
+    public boolean registerCustomItem(@NotNull String identifier, @NotNull CustomItemData customItemData) {
+        if (this.cantRegisterNewItem()) {
+            return false;
+        }
+
+        GeyserCustomMappingData registeredItem = CustomItemRegistryPopulator.populateRegistry(identifier, customItemData, this.registeredItemCount());
+
+        if (registeredItem.mappingNumber() > 0) {
+            this.customMappings.computeIfAbsent(identifier, list -> new ArrayList<>()).add(customItemData);
+            this.registeredCustomItems.add(registeredItem);
+
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean registerCustomItem(@NonNull NonVanillaCustomItemData customItemData) {
+        if (this.cantRegisterNewItem()) {
+            return false;
+        }
+
+        Int2ObjectMap<ComponentItemData> componentItemData = CustomItemRegistryPopulator.populateRegistry(customItemData, this.registeredItemCount());
```

Comment:
See the other `registerCustomItem` method, but this one is worse since this one loops through every single item.